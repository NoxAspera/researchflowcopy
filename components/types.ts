/**
 * Types.ts
 * @author Blake Stambaugh
 * Updated: 1/14/25 - MO
 * 
 * Helper that will contain more complicated types, like those for navigation.
 */
import { StackNavigationProp } from '@react-navigation/stack';
import { RouteProp } from '@react-navigation/native';
import { visit } from '../scripts/APIRequests';

// Define the type for the stack's navigation parameters
export type RootStackParamList = {
    // used for home screen
    SelectSite: { from: string }; 
    SelectInstrument: { from: string , notes: boolean};
    SelectTank: { from: string, onSelect?: (selectedTank: string) => void };
    SelectNotes: { from: string }; 

    // used for site select
    AddNotes: { site: string };
    AddNotesMobile: { site: string };
    ViewNotes: { site: string, from: string, visits: visit[] };
    BadData: { site: string };
    InstrumentMaintenance: { site: string };
    TankTracker: { site: string };
    PlanVisit: {site: string};
    Diagnostics: {site: string};
    ContactInfo: {site: string};

    // used for auth
    Home: undefined;
    ViewNotifications: undefined;
    NotificationsButton: undefined;
    Calendar: undefined;
};

// Type for the navigation prop for the HomeScreen
export type NaviProp = StackNavigationProp<RootStackParamList, 'SelectSite'>;

// Props for the HomeScreen component
export interface NavigationType {
  navigation: NaviProp;
}

// type for routes
type RouteParams = {
  site: string; 
  info: string; 
  from: string;
  notes: string;
  onSelect?: (selectedTank: string ) => void;
  visits: visit[];
};
export type routeProp = RouteProp<{params: RouteParams}, 'params'>;