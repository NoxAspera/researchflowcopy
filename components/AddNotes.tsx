/**
 * Add Notes Page
 * @author Blake Stambaugh, Megan Ostlie, August O'Rourke, and David Schiwal
 *  12/4/24
 * This page will take in input from the user, format it, and upload it to the
 * github repo.
 */
import { StyleSheet, KeyboardAvoidingView, Platform } from 'react-native';
import React, { useState, useEffect } from 'react';
import { RouteProp, useRoute } from '@react-navigation/native';
import { ScrollView } from 'react-native-gesture-handler';
import { buildNotes, Entry } from '../scripts/Parsers';
import { NaviProp } from './types';
import TextInput from './TextInput'
import NoteInput from './NoteInput'
import { IndexPath, Layout, Select, SelectItem, Button, Text } from '@ui-kitten/components';
import * as eva from '@eva-design/eva';
import { customTheme } from './CustomTheme'
import { setFile, getFileContents } from '../scripts/APIRequests';
import { parseNotes, ParsedData } from '../scripts/Parsers'
import PopupProp from './Popup';
import PopupProp2Button from './Popup2Button';
import { NavigationType, routeProp } from './types'
import { ThemeContext } from './ThemeContext';


function checkValidTime(entry:string)
{
  const timeMatch = /^[0-2][0-9]:[0-5][0-9]$/gm
  return timeMatch.test(entry)
}

function checkValidNumber(entry:string)
{
  const anyNonNumber = /^(\d+)(\.\d+)?$/gm
 return anyNonNumber.test(entry)
}

/**
 * @author Megan Ostlie
 *  a function that pulls the current note document for the specified site from GitHub
 *  @param siteName the name of the site
 * 
 * @returns a ParsedData object that contains the information of the given document
 */
async function processNotes(siteName: string) {
  const fileContents = await getFileContents(`site_notes/${siteName}`);
  if(fileContents.data){
    return parseNotes(fileContents.data)
  }
  else
  {
    return null
  }
}

/**
 * @author August O'Rourke, Blake Stambaugh, David Schiwal, Megan Ostlie
 *  Creates the input elements for the user to input site note information.
 *  Pulls the current notes for the selected site from GitHub and autofills certain fields.
 *  Takes the inputted information from the user to build a new string that is added to that site's note document.
 * 
 */
export default function AddNotes({ navigation }: NavigationType) {
    const route = useRoute<routeProp>();
    const { site, info } = route.params || {}
    const themeContext = React.useContext(ThemeContext);
    const isDarkMode = themeContext.theme === 'dark';

    // State to hold parsed data
    const [data, setData] = useState<ParsedData | null>(null);

    // Get current notes for the site
    useEffect(() => {
        async function fetchData() {
            if (site && !data) {
                try {
                    const parsedData = await processNotes(site);
                    setData(parsedData); // Update state with the latest entry
                } catch (error) {
                    console.error("Error processing notes:", error);
                }
            }
        }
        fetchData();
    }, [site]); // Re-run if `site` changes

    //Get latest notes entry from site
    let latestEntry = null;
    if (data) {
      latestEntry = data.entries[0];
    }
    
    // these use states to set and store values in the text inputs
    const [timeValue, setTimeValue] = useState("");
    const [nameValue, setNameValue] = useState("");
    const [ltsId, setLTSId] = useState("");
    const [ltsValue, setLTSValue] = useState("");
    const [ltsPressure, setLTSPressure] = useState("");
    const [lowId, setLowId] = useState("");
    const [lowValue, setLowValue] = useState("");
    const [lowPressure, setLowPressure] = useState("");
    const [midId, setmidId] = useState("");
    const [midValue, setmidValue] = useState("");
    const [midPressure, setmidPressure] = useState("");
    const [highId, setHighId] = useState("");
    const [highValue, setHighValue] = useState("");
    const [highPressure, setHighPressure] = useState("");
    const [n2Value, setN2Value] = useState("");
    const [notesValue, setNotesValue] = useState("");
    const [instrumentInput, setInstrumentInput] = useState("");

    // Use IndexPath for selected index for drop down menu
    const [selectedIndex, setSelectedIndex] = useState<IndexPath>(new IndexPath(0)); // Default to first item
    const [instruments, setInstruments] = useState<string[]>(['Instrument 1']);

    // used for determining if PUT request was successful
    // will set the success/fail notification to visible, aswell as the color and text
    const [visible, setVisible] = useState(false);
    const [messageColor, setMessageColor] = useState("");
    const [message, setMessage] = useState("");
    //used for popup if info is missing
    const [visible2, setVisible2] = useState(false);

    //method will warn user if fields haven't been input
    function checkTextEntries(){
        if(timeValue == "" ||
           nameValue == "" ||
           ltsId == "" ||
           ltsValue == "" ||
           ltsPressure == "" ||
           lowId == "" ||
           lowValue == "" ||
           lowPressure == "" ||
           midId == "" ||
           midValue == "" ||
           midPressure == "" ||
           highId == "" ||
           highValue == "" ||
           highPressure == "" ||
           n2Value == "" ||
           notesValue == ""){
               setVisible2(true);
        }
        else{
            handleUpdate();
           }
    }

    // will call setFile to send the PUT request. 
    // If it is successful it will display a success message
    // if it fails then it will display a failure message
    const handleUpdate = async () => {
        const LTSignored: boolean = (ltsId == "" && ltsValue == "" && ltsPressure == "")

        // get time submitted
        const now = new Date();
        const year = now.getFullYear().toString()
        const month = (now.getMonth() + 1).toString() // now.getMonth() is zero-base (i.e. January is 0), likely due to something with Oracle's implementation - August
        const day = now.getDate().toString()
        const hours= now.getHours().toString()
        const minutes = now.getMinutes().toString()
        
        // create an entry object data that will be sent off to the repo
        let data: Entry = 
        {
          time_in: `${year}-${month}-${day} ${timeValue}`,
          time_out: `${year}-${month}-${day} ${hours}:${minutes}`,
          names: nameValue,
          instrument: instrumentInput.trim() ? instrumentInput : (instruments[selectedIndex.row] ? instruments[selectedIndex.row] : null),
          n2_pressure: n2Value ? n2Value: null,
          lts: LTSignored ? null:
          {
            id: ltsId,
            value: ltsValue,
            unit: "ppm",
            pressure: ltsPressure
          },
          low_cal:
          {
            id:lowId,
            value:lowValue,
            unit:"ppm",
            pressure: lowPressure
          },
          mid_cal:
          {
            id:midId,
            value:midValue,
            unit: "ppm",
            pressure: midPressure
          },
          high_cal:
          {
            id:highId,
            value:highValue,
            unit: "ppm",
            pressure: highPressure
          },
          additional_notes: notesValue 
        };

        // send the request
        const result = await setFile(site, buildNotes(data), "updating notes from researchFlow");

        // if the warning popup is visible, remove it
        //if(visible2) { setVisible2(false); }

        // check to see if the request was ok, give a message based on that
        if (result.success) {
            setMessage("File updated successfully!");
            navigation.navigate("Home")
            setMessageColor(customTheme['color-success-700']);
          } else {
            setMessage(`Error: ${result.error}`);
            setMessageColor(customTheme['color-danger-700']);
          }
        setVisible(true);
    };

    //Set tank ids, values, and instruments if available in parsed data
    useEffect(() => {
      if (latestEntry) {
          if (latestEntry.lts) {
              setLTSId(latestEntry.lts.id || "");
              setLTSValue(latestEntry.lts.value || "");
          }
          if (latestEntry.low_cal) {
            setLowId(latestEntry.low_cal.id || "");
            setLowValue(latestEntry.low_cal.value || "");
          }
          if (latestEntry.mid_cal) {
            setmidId(latestEntry.mid_cal.id || "");
            setmidValue(latestEntry.mid_cal.value || "");
          }
          if (latestEntry.high_cal) {
            setHighId(latestEntry.high_cal.id || "");
            setHighValue(latestEntry.high_cal.value || "");
          }
          if (latestEntry.instrument) {
            setInstrumentInput(latestEntry.instrument);
          }
      }
  }, [latestEntry]);

    return (
      <KeyboardAvoidingView
            behavior = "padding"
            style={styles.container}>
        <ScrollView automaticallyAdjustKeyboardInsets={true}>
          <Layout style={styles.container}>
            {/* header */}
            <Text category='h1' style={{textAlign: 'center'}}>{site}</Text>

            {/* success/failure popup */}
            <PopupProp popupText={message}
            popupColor={messageColor} 
            onPress={setVisible} 
            visible={visible}/>

            {/* popup if user has missing input */}
            <PopupProp2Button popupText='Missing some input field(s)'
            popupColor={customTheme['color-danger-700']}
            sendData={handleUpdate}
            removePopup={setVisible2}
            visible={visible2}/>

            {/* text box for instrument */}
              <TextInput
                labelText="Instrument"
                labelValue={instrumentInput}
                onTextChange={setInstrumentInput}
                placeholder="Please enter instrument name"
                style={styles.inputText}
              />

            {/* text inputs */}
            {/* Name input */}
            <TextInput labelText='Name' 
              labelValue={nameValue} 
              onTextChange={setNameValue} 
              placeholder='ResearchFlow' 
              style={styles.inputText}
            />

              {/* Time input */}
            <TextInput labelText='Time Started' 
              labelValue={timeValue} 
              onTextChange={setTimeValue} 
              placeholder='15:00' 
              style={styles.inputText} />

            {/* N2 */}
            <TextInput labelText='N2 (if needed)' 
              labelValue={n2Value} 
              onTextChange={setN2Value} 
              placeholder='Pressure' 
              style={styles.inputText} />

            {/* LTS input */}
            <Layout style = {styles.rowContainer}>
              <TextInput labelText='LTS (if needed)' 
                labelValue={ltsId} 
                onTextChange={setLTSId} 
                placeholder='Tank ID' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={ltsValue} 
                onTextChange={setLTSValue} 
                placeholder='Value' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={ltsPressure} 
                onTextChange={setLTSPressure} 
                placeholder='Pressure' 
                style={styles.tankInput} />
            </Layout>

            {/* Note for all tank inputs, the single space labels are there to make sure the other entry fields are alligned well*/}

            {/* Low input */}
            <Layout style = {styles.rowContainer}>
              <TextInput labelText='Low' 
                labelValue={lowId} 
                onTextChange={setLowId} 
                placeholder='Tank ID' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={lowValue} 
                onTextChange={setLowValue} 
                placeholder='Value' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={lowPressure} 
                onTextChange={setLowPressure} 
                placeholder='Pressure' 
                style={styles.tankInput} />
            </Layout>

            {/* mid input */}
            <Layout style = {styles.rowContainer}>
              <TextInput labelText='Mid' 
                labelValue={midId} 
                onTextChange={setmidId} 
                placeholder='Tank ID' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={midValue} 
                onTextChange={setmidValue} 
                placeholder='Value' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={midPressure} 
                onTextChange={setmidPressure} 
                placeholder='Pressure' 
                style={styles.tankInput} />
            </Layout>

            {/* high input */} 
            <Layout style = {styles.rowContainer}>
              <TextInput labelText='High' 
                labelValue={highId} 
                onTextChange={setHighId} 
                placeholder='Tank ID' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={highValue} 
                onTextChange={setHighValue} 
                placeholder='Value' 
                style={styles.tankInput} />
              <TextInput labelText=' ' 
                labelValue={highPressure} 
                onTextChange={setHighPressure} 
                placeholder='Pressure' 
                style={styles.tankInput} />
            </Layout>

            {/* notes entry */}
            <NoteInput labelText='Notes' 
              labelValue={notesValue} 
              onTextChange={setNotesValue} 
              placeholder='All Good.' 
              multiplelines={true} 
              style={styles.notesInput}/>

            {/* submit button */}
            <Button
              testID='AddNotesButton'
              onPress={() => checkTextEntries()}
              appearance='filled'
              status='primary' 
              style={styles.submitButton}>
              {evaProps => <Text {...evaProps} category="h6" style={{color: "black"}}>Submit</Text>}
            </Button>
          </Layout>
        </ScrollView>
      </KeyboardAvoidingView>
    );
  }
  
  const styles = StyleSheet.create({
    container: {
      flex: 1,
      alignItems: 'stretch',        // has button fill space horizontally
      justifyContent: 'space-evenly',
    },
    inputText: {
      flex: 1,
      margin: 8
    },
    tankInput: {
      flex: 1,
      margin: 8
    },
    notesInput:
    {
      flex: 1,
      margin: 8,
    },
    areaInput: {
      height: 200,
      margin: 10,
      width: 380,
      borderWidth: 1,
      padding: 10,
    },
    rowContainer:
    {
      flex: 1,
      flexDirection: 'row',
      alignItems: 'stretch',        // has button fill space horizontally
      justifyContent: 'space-evenly',
    },
    submitButton:
    {
      margin: 20, 
      backgroundColor: "#06b4e0",
    },
});