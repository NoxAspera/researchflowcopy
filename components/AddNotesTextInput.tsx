import { View, Text, StyleSheet, ViewStyle } from "react-native";
import { TextInput } from "react-native-gesture-handler";
import { SafeAreaView, SafeAreaProvider } from "react-native-safe-area-context";

interface ANTIProp{
    inputText: string;
    inputTextVar?: string;
    inputTextStyle: ViewStyle;
    ViewStyle: ViewStyle; 
}

const AddNotesTextInput: React.FC<ANTIProp> = ({ inputText, inputTextStyle, ViewStyle, inputTextVar }) => {
    return (
        <View style = {ViewStyle}>
            <Text style = {styles.label}>{inputText}{inputTextVar}</Text>
            <SafeAreaProvider>
              <SafeAreaView>
                <TextInput
                  style = {inputTextStyle}>
                </TextInput>
              </SafeAreaView>
            </SafeAreaProvider>
          </View>
    );
};

const styles = StyleSheet.create({
    label: {
        margin: 15,
        fontSize: 24,
        alignItems: 'flex-start'
    }
  });

  export default AddNotesTextInput;