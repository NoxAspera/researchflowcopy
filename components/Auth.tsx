import { useEffect, useState } from 'react';
import * as WebBrowser from 'expo-web-browser';
import * as config from "../config"
import { makeRedirectUri, useAuthRequest, AuthSessionResult} from 'expo-auth-session';
import { generateGithubToken, readUpdates, setGithubToken, tankTrackerOffline, tankTrackerSpinUp, updateDirectories } from '../scripts/APIRequests';
import { StyleSheet, ScrollView} from 'react-native';
import { NavigationType} from './types'
import React from 'react';
import {Layout} from '@ui-kitten/components'
import HomeButtonProp from './HomeButtonProp';
import * as Network from 'expo-network'
import LoadingScreen from './LoadingScreen';


WebBrowser.maybeCompleteAuthSession();

// Endpoint
const discovery = {
  authorizationEndpoint: 'https://github.com/login/oauth/authorize',
  tokenEndpoint: 'https://github.com/login/oauth/access_token',
  revocationEndpoint: 'https://github.com/settings/connections/applications/Iv23liZ9ogrXCPdG093f',
};


async function isConnected()
{
  let check = (await Network.getNetworkStateAsync()).isConnected
  return check
}

export default function App({navigation}: NavigationType) {
  const [networkStatus, setNetworkStatus] = useState(true)
  const [loadingValue, setLoadingValue] = useState(false);

  console.log(makeRedirectUri({scheme: 'researchflowuofu'}))
  const [request, response, promptAsync] = useAuthRequest(
    {
      clientId: config.GITHUB_CLIENT_ID,
      scopes: ['repo'],
      redirectUri: makeRedirectUri({
        scheme: 'researchflowuofu'
      }),
    },
    discovery
  );

  async function handleResponse(response: AuthSessionResult | null)
  {
    if (response?.type === 'success') {
      setLoadingValue(true)
      console.log(response)
        const { code } = response.params;
        
        const {token_type, scopes, access_token} = (await generateGithubToken(code)).data
        setGithubToken(access_token);
        await tankTrackerSpinUp();
        await readUpdates()
        await updateDirectories()
        setLoadingValue(false)
    }
  }

  useEffect(() => {
    handleResponse(response)
  }, [response]);

async function handlePress()
{
  setNetworkStatus(await isConnected())
  console.log(networkStatus)
  if(networkStatus)
  {
    await promptAsync();
  }
  else
  {
    setLoadingValue(true)
    await tankTrackerOffline()
    setLoadingValue(false)
  }
  navigation.navigate("Home")
}

  return (
    <Layout style={styles.container}>
      <ScrollView contentContainerStyle={styles.scrollContainer}>
      <LoadingScreen visible={loadingValue} />
        <HomeButtonProp
          text="Login With Github"
          color= "#FFFFFF"
          onPress={() => {
            handlePress()
          }}
        />
        </ScrollView>
    </Layout>
      
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',        // has button fill space horizontally
    justifyContent: 'space-evenly',
  },
  scrollContainer: {
    paddingVertical: 16,
    alignItems: 'center', // Center cards horizontally
  },
});
