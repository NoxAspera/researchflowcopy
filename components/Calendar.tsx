/**
 * Home Screen
 * @author Blake Stambaugh
 * 11/26/2024
 *
 * The follow code represents the home page the user sees when they first launch our app.
 * It has a button for each section of the app that will take them to the next page.
 **/
import { StyleSheet, Dimensions } from "react-native";
import HomeButtonProp from "./HomeButtonProp";
import { Layout } from "@ui-kitten/components";
import React, {useEffect, useState} from "react";
import { NavigationType } from "./types";
const { width, height } = Dimensions.get("window"); //this pulls in the screen width and height to use for scalars
import * as calendar from 'expo-calendar';
import { getFileContents, visit } from "../scripts/APIRequests";
import { processVisits } from "../scripts/Parsers";
import { DateData, LocaleConfig} from 'react-native-calendars'
import { Calendar, CalendarList,Agenda } from "react-native-calendars";
import { customTheme } from './CustomTheme';
import PopupProp from './Popup';
import PlanVisit from "./PlanVisit";

let visitDict: Map<Date, visit>

function handlePress(date: DateData)
{

}

export default function CalendarScreen({ navigation }: NavigationType) {
  const [markedDates, setMarkedDates] = useState({});
  const [visible, setVisible] = useState(false);
  const [messageColor, setMessageColor] = useState("");
  const [message, setMessage] = useState("");

  useEffect(() => {
    async function fetchData() {
      try{
        let response = await getFileContents("researchflow_data/visits")
        if (response.success)
        {
          visitDict = new Map()
          let visits: visit[] = processVisits(response.data)
          visits.forEach((value) => {
            console.log(value.date)
            let date: Date = new Date (Date.parse(value.date))
            visitDict.set(date, value)
            if(markedDates)
            {
              setMarkedDates(prevmarkedDates => ({...prevmarkedDates,[value.date] :{marked: true, dotColor: 'blue'}}))
            }
          })
        }
        else
        {
          setMessage(`Error: ${response.error}`);
          setMessageColor(customTheme["color-danger-700"]);
          setVisible(true);
        }
      }
      catch(error)
      {
        console.log(error)
      }
    }
    fetchData();
}, []);

  return (
    <Layout style={styles.container}>
        {/*If you add a new navigation point you have to add it in types.ts, and App.tsx*/}
        <PopupProp
          popupText={message}
          popupColor={messageColor}
          onPress={setVisible}
          visible={visible}
        />
        {/* Plan a Visit */}
        <HomeButtonProp
          text="PLAN A VISIT"
          color="#4DD7FA"
          onPress={() =>
            navigation.navigate("SelectSite", {from: "PlanVisit"})
          }
        />
        <Calendar
          onDayPress={Date => {handlePress(Date)}}
          markedDates={markedDates}
        />
    </Layout>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center", // has button fill space horizontally
    justifyContent: "space-evenly",
  },
  scrollContainer: {
    paddingVertical: 16,
    alignItems: "center", // Center cards horizontally
  },
});
