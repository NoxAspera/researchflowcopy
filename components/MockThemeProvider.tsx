import React, { useState } from "react";
import { ThemeContext } from "./ThemeContext";
import { ApplicationProvider, IconRegistry } from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import { EvaIconsPack } from '@ui-kitten/eva-icons';
import customColors from '../custom-theme.json'

export const MockThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState('light');

  const currentTheme = { ...eva[theme], ...customColors };

  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme === 'light' ? 'dark' : 'light'));
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <IconRegistry icons={EvaIconsPack} />
      <ApplicationProvider {...eva} theme={currentTheme}>
        {children}
      </ApplicationProvider>
    </ThemeContext.Provider>
  );
};