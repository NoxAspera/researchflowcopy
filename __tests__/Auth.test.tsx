import Auth from '../components/Auth';
import { fireEvent, render } from '@testing-library/react-native'
import { NaviProp } from '../components/types';
import React from 'react';
import { MockThemeProvider } from '../components/MockThemeProvider';



beforeEach(() => {
  jest.spyOn(console, 'error')
  // @ts-ignore jest.spyOn adds this functionallity
  console.error.mockImplementation(() => null);
});

afterEach(() => {
  // @ts-ignore jest.spyOn adds this functionallity
  console.error.mockRestore()
})

const mockNavigation: NaviProp = {
  dispatch: jest.fn(),
  navigate: jest.fn(),
  navigateDeprecated: jest.fn(),
  preload: jest.fn(),
  reset: jest.fn(),
  goBack: jest.fn(),
  isFocused: jest.fn(),
  canGoBack: jest.fn(),
  getId: jest.fn(),
  getState: jest.fn(),
  setStateForNextRouteNamesChange: jest.fn(),
  getParent: jest.fn(),
  setParams: jest.fn(),
  setOptions: jest.fn(),
  addListener: jest.fn(),
  removeListener: jest.fn(),
  replace: jest.fn(), 
  push: jest.fn(), 
  pop: jest.fn(), 
  popToTop: jest.fn(), 
  popTo: jest.fn()
}

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
  useRoute: () => ({
    params: {},
  }),
}));

jest.mock('react-native-gesture-handler', () => {
  const { View } = require('react-native');
  return {
    Swipeable: View,
    DrawerLayout: View,
    State: {},
    ScrollView: View,
    Slider: View,
    Switch: View,
    TextInput: View,
    ToolbarAndroid: View,
    ViewPagerAndroid: View,
    WebView: View,
    NativeViewGestureHandler: View,
    TapGestureHandler: View,
    FlingGestureHandler: View,
    ForceTouchGestureHandler: View,
    LongPressGestureHandler: View,
    PanGestureHandler: View,
    PinchGestureHandler: View,
    RotationGestureHandler: View,
    /* Buttons */
    RawButton: View,
    BaseButton: View,
    RectButton: View,
    BorderlessButton: View,
    /* Others */
    FlatList: View,
    gestureHandlerRootHOC: jest.fn(),
    Directions: {},
  };
});

describe('Login page', () => {
  test('page renders correctly', () => {
    const { toJSON } = render(
    <MockThemeProvider>
      <Auth navigation={mockNavigation} />
    </MockThemeProvider>
  );
  expect(toJSON()).toMatchSnapshot();
  });

  test('can login and have the page change', () => {
    const { getByTestId, queryByPlaceholderText } = render(
      <MockThemeProvider>
        <Auth navigation={mockNavigation} />
      </MockThemeProvider>
    );

    // get inputs and button
    const email = queryByPlaceholderText('Email');
    const pass = queryByPlaceholderText('Password');
    const button = getByTestId('AuthButton');
    
    // give input
    fireEvent.changeText(email, 'admin');
    fireEvent.changeText(pass, '1234');

    // hit button to check for success
    fireEvent.press(button);
    expect(mockNavigation.navigate).toHaveBeenCalledWith('Home');
  })

  test('does not redirect with bad login', () => {
    const { getByTestId, getByText } = render(
      <MockThemeProvider>
        <Auth navigation={mockNavigation} />
      </MockThemeProvider>
    );

    // get button
    const button = getByTestId('AuthButton');

    // hit button to check for failure
    fireEvent.press(button);
    const popup = getByText("Missing Login Credentials")
    expect(popup).toBeTruthy();
  })

})