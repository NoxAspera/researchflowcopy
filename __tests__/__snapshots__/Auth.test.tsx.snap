// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page page renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "alignItems": "stretch",
        "flex": 1,
        "justifyContent": "space-evenly",
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    appearance="default"
    level="1"
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "alignItems": "stretch",
          "flex": 1,
          "justifyContent": "space-evenly",
        },
      ]
    }
  >
    <View
      appearance="default"
      level="1"
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "alignItems": "stretch",
            "backgroundColor": "#C8D6F6",
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        appearance="default"
        category="h1"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 36,
              "fontWeight": "800",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Hello
      </Text>
      <Text
        appearance="default"
        category="h6"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 18,
              "fontWeight": "800",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Sign in using your GitHub credentials
      </Text>
    </View>
    <View
      appearance="default"
      level="1"
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
          },
          undefined,
        ]
      }
      testID=""
    />
    <View
      appearance="default"
      level="1"
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "flex": 3,
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "margin": 15,
            "opacity": 1,
          }
        }
        testID="@/container"
      >
        <Text
          appearance="default"
          category="p2"
          status="basic"
          style={
            [
              {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 13,
                "fontWeight": "400",
              },
              {
                "color": "black",
              },
            ]
          }
        />
        <View
          style={
            [
              {
                "backgroundColor": "#F7F9FC",
                "borderColor": "#E4E9F2",
                "borderRadius": 4,
                "borderWidth": 1,
                "minHeight": 40,
                "paddingHorizontal": 8,
                "paddingVertical": 7,
              },
              {
                "alignItems": "center",
                "flexDirection": "row",
                "width": "100%",
              },
            ]
          }
        >
          <TextInput
            appearance="default"
            editable={true}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            placeholder="Email"
            placeholderTextColor="#A0A0A0"
            secureTextEntry={false}
            size="medium"
            status="basic"
            style={
              [
                {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "normal",
                  "marginHorizontal": 8,
                },
                {
                  "flexBasis": "auto",
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
                null,
                {
                  "fontSize": 18,
                  "maxHeight": 24,
                },
              ]
            }
            testID="@/input"
            value=""
          />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "margin": 15,
            "opacity": 1,
          }
        }
        testID="@/container"
      >
        <Text
          appearance="default"
          category="p2"
          status="basic"
          style={
            [
              {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 13,
                "fontWeight": "400",
              },
              {
                "color": "black",
              },
            ]
          }
        />
        <View
          style={
            [
              {
                "backgroundColor": "#F7F9FC",
                "borderColor": "#E4E9F2",
                "borderRadius": 4,
                "borderWidth": 1,
                "minHeight": 40,
                "paddingHorizontal": 8,
                "paddingVertical": 7,
              },
              {
                "alignItems": "center",
                "flexDirection": "row",
                "width": "100%",
              },
            ]
          }
        >
          <TextInput
            appearance="default"
            editable={true}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            placeholder="Password"
            placeholderTextColor="#A0A0A0"
            secureTextEntry={true}
            size="medium"
            status="basic"
            style={
              [
                {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "normal",
                  "marginHorizontal": 8,
                },
                {
                  "flexBasis": "auto",
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
                null,
                {
                  "fontSize": 18,
                  "maxHeight": 24,
                },
              ]
            }
            testID="@/input"
            value=""
          />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#06b4e0",
            "borderColor": "#596CA7",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 15,
            "minHeight": 40,
            "minWidth": 40,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
          }
        }
        testID="AuthButton"
      >
        <Text
          appearance="default"
          category="h5"
          status="basic"
          style={
            [
              {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 22,
                "fontWeight": "800",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          SIGN IN
        </Text>
      </View>
    </View>
  </View>
</View>
`;
